---
title: "Single Cell RNA-seq Basic Analysis Pipeline with Seurat"
author: "Zain Arifin"
date: "2022-08-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(SeuratData)
library(patchwork)
library(tidyverse)
```

## Single Cell Analysis

```{r load-data, warning=F}
InstallData("ifnb")
LoadData("ifnb")

# In this dataset, the source of batch is stimulation condition
ifnb.list <- SplitObject(ifnb, split.by = "stim")
```

## Quality control

```{r}
for (i in 1:length(ifnb.list)){
  ifnb.list[[i]][["percent.mt"]] <- PercentageFeatureSet(ifnb.list[[i]], pattern = "^MT-")
  print(VlnPlot(ifnb.list[[i]], features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3))
}
#Data has been QCed
```


## Preprocess scRNA-seq

Prior integration, dataset must be pre processed independently. The steps are:

1. Normalization
The fastest and easiest way to normalize scRNA-seq data is to divided the raw value by total individual raw count followed log-transformation.

2. Find variable features
Calculate variance for each gene, sort them, and take top 2,000 most variables genes (default is 2,000).

```{r preprocess}
ifnb.list <- lapply(X = ifnb.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 3000)
})
features <- SelectIntegrationFeatures(object.list = ifnb.list, verbose = F)
```


## Integration

3. Batch correction/integration
Seurat uses Canonical Correlation Analysis (CCA) to find "anchors" to merge scRNA-seq data. Anchors refer to a similar biological state measured and weighted based on overlap in nearest neighbors. Seurat transforms the dataset into a shared space which allow mapping of of new data to the estimated referrence.


```{r integration}
anchors <- FindIntegrationAnchors(object.list = ifnb.list, anchor.features = features, verbose = F)
dat.combined <- IntegrateData(anchorset = anchors) # create integrated seurat object
DefaultAssay(dat.combined) <- 'integrated'
dat.combined <- ScaleData(dat.combined, verbose = FALSE)
dat.combined <- RunPCA(dat.combined, npcs = 10, verbose = FALSE)
dat.combined <- FindNeighbors(dat.combined, dims = 1:10)
dat.combined <- RunUMAP(dat.combined, dims = 1:10)
dat.combined <- FindClusters(dat.combined, resolution = 0.7) 
```

```{r viz}
p1 <- DimPlot(dat.combined, group.by = "seurat_clusters", label = T, repel = T)
p2 <- DimPlot(dat.combined, group.by = "seurat_annotations", label = T, repel = T)
p3 <- DimPlot(dat.combined, group.by = "stim")

p1
p2
p3
```

## DEG Analysis

Performing DEG can be useful to find differentially expressed genes that can be used as a marker. By default, Seurat uses Wilcoxon rank sum test. The code below find markers for all cell type and return only positive markers

```{r}
dat.combined <- SetIdent(dat.combined, value = "seurat_annotations")
markers_all <- FindAllMarkers(dat.combined, only.pos = TRUE, min.pct = 0.25, 
                              logfc.threshold = 0.25)
top_markers <- markers_all %>%
    group_by(cluster) %>%
    slice_max(n = 1, order_by = avg_log2FC)
```

Visualization of DEGs 

```{r}
p2
for (i in 1:nrow(top_markers)){
  print(FeaturePlot(dat.combined, features = as.vector(top_markers$gene[i])))
}

markers_all %>%
    group_by(cluster) %>%
    top_n(n = 5, wt = avg_log2FC) -> top5
DoHeatmap(dat.combined, features = top5$gene) + NoLegend()
```
